File: ProblemSet_25xx/Problem_2561/solution.py
# Write your solution here
from typing import Counter, List


class Solution:
    def minCost(self, basket1: List[int], basket2: List[int]) -> int:
        basket1_counter = Counter(basket1)
        basket2_counter = Counter(basket2)
        counter = basket1_counter + basket2_counter
        
        swap = []
        min_key = min(counter.keys())
        for key in counter.keys():
            if (counter[key] % 2 != 0):
                return -1
            freq = counter[key] // 2
            diff = freq - basket1_counter[key]
            swap += [key] * abs(diff)
        swap.sort()
        
        result = 0
        for value in swap[0:len(swap)//2]:
            if value > min_key * 2:
                result += 2 * min_key
            else:
                result += value
        return result
    
s = Solution()
print('Result', s.minCost(
    [84,80,43,8,80,88,43,14,100,88],
    [32,32,42,68,68,100,42,84,14,8]))

File: ProblemSet_25xx/Problem_2561/solution.cs
[TestClass]
public class Solution
{
    [TestMethod]
    public void MinCostTest()
    {
        Console.WriteLine(MinCost([4, 2, 2, 2], [1, 4, 1, 2]));
        Console.WriteLine(MinCost([2, 3, 4, 1], [3, 2, 5, 1]));
        Console.WriteLine(MinCost([4, 4, 4, 4, 3], [5, 5, 5, 5, 3]));
    }
    public long MinCost(int[] basket1, int[] basket2)
    {
        List<int> needtomovefromb1 = [];
        List<int> needtomovefromb2 = [];
        Array.Sort(basket1);
        Array.Sort(basket2);
        int min = Math.Min(basket1[0], basket2[0]);
        int ptr1 = 0; int ptr2 = 0;
        while (ptr1 < basket1.Length && ptr2 < basket1.Length)
        {
            if (basket1[ptr1] > basket2[ptr2])
            {
                if (ptr2 + 1 < basket1.Length && basket2[ptr2++] == basket2[ptr2++])
                    needtomovefromb2.Add(basket2[ptr2 - 2]);
                else return -1;
            }
            else if (basket1[ptr1] == basket2[ptr2])
            {
                ptr1++; ptr2++;
            }
            else
            {
                if (ptr1 + 1 < basket1.Length && basket1[ptr1++] == basket1[ptr1++])
                    needtomovefromb1.Add(basket1[ptr1 - 2]);
                else return -1;
            }
        }
        while (ptr1 < basket1.Length)
        {
            if (ptr1 + 1 < basket1.Length && basket1[ptr1++] == basket1[ptr1++])
                needtomovefromb1.Add(basket1[ptr1 - 2]);
            else return -1;
        }
        while (ptr2 < basket1.Length)
        {
            if (ptr2 + 1 < basket1.Length && basket2[ptr2++] == basket2[ptr2++])
                needtomovefromb2.Add(basket2[ptr1 - 2]);
            else return -1;
        }
        int score = 0;
        for (int i = 0; i < needtomovefromb1.Count; i++)
        {
            int value = Math.Min(needtomovefromb1[i], needtomovefromb2[needtomovefromb1.Count - i - 1]);
            if (value > min * 2) value = min * 2;
            score += value;
        }
        return score;
    }
}